Fixes ENYO-1641

ryanjduffy opened the pull request:
## Issue

We have an existing and 2 in-flight overlay implementations that need to be consolidated into a consistent API.
## Fix

Add samples for moonstone/Overlay

Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

Commit:
# This is a combination of 5 commits.
# The first commit's message is:
add Overlay sample

Also updates other samples using overlays

Issue: ENYO-1641
Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

# This is the 2nd commit message:

rename overlay sample and fix up related samples

Issue: ENYO-1641
Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

# This is the 3rd commit message:

add ClampedText example to OverlaySample

Issue: ENYO-1641
Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

# This is the 4th commit message:

convert ImageBadgeSample to new overlay design

Issue: ENYO-1641
Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

# This is the 5th commit message:

add placeholder for future support

ENYO-510 will add placeholder support for images so preemptively adding
placeholders in the sample. Also added fixed sizes for the images so
the overlays appear in the right place while loading

Issue: ENYO-1641
Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

Commit:
add Overlay sample

Also updates other samples using overlays
rename overlay sample and fix up related samples
convert ImageBadgeSample to new overlay design

ENYO-510 will add placeholder support for images so preemptively adding
placeholders in the sample. Also added fixed sizes for the images so
the overlays appear in the right place while loading
use IconButtons in OverlaySample

Issue: ENYO-1641
Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

Djspaceg commented on src/moonstone-samples/lib/OverlaySample.js:
```diff
@@ -0,0 +1,145 @@
+					// visible, clamped must be set to 'always' because the dimensions cannot be
+					// accurately calculated when the ClampedText control isn't showing.
+					{kind: ClampedText, maxLines: 3, clamped: 'always', classes: 'moon-overlay-text-title', content: 'This is probably too long to put in a text overlay but let\'s say you wanted to anyway'},
+					{classes: 'moon-overlay-text-subtitle', content: '3 lines'},
```
Trailing comma. (Last one, I promise)

ryanjduffy replied:
trailing comma ... gumble grumble ... blake ... grumble ...

Djspaceg replied:
I approved it before this comma was added. ;) I just figure it's better to catch them now than to have to go back later.

aarontam commented:
Hi @ryanjduffy, any chance you can fix the merge conflict sometime today? If not, I'll take care of it!

aarontam commented:
As a general question/comment, I noticed with the samples that involve loading an image from a remote URL (and with a somewhat slow connection right now), the delay in loading the image causes the overlay to be incorrectly positioned initially, until the entire image is loaded. I think we might want to consider more gracefully handling these cases as slow-loading thumbnails or images can be a common scenario. Thoughts on this?

Djspaceg commented:
Sounds like a good use of our shiny new image placeholder support. :D 

But in general, Images without a height and width assigned will behave as you describe; the browser automatically resizes them post loading. Without knowing their final dimensions before we start loading them, I don't think there's much/anything we can do differently. That being said, in the TV scenario, this will be less common because we have local assets and/or the sizes are handled by our GridListImageItem which assigns the sizes itself. It's possible, but unlikely in production because there's preventative measures apps can take to avoid it.

aarontam commented:
Yep that sounds good to me. :) I was thinking more about the scenario where developers will be copying this code verbatim - if we had some generally best practices for these cases, such as using the placeholder image (when it's merged) or specifying explicit dimensions, then we would at least cover these cases sufficiently.

Edit: Ah I see @ryanjduffy added those, thanks!

aarontam committed:
ENYO-1641: Update sample to properly reflect marginless moon.Image inside moon.GridListImageItem.

Enyo-DCO-1.1-Signed-off-by: Aaron Tam <aaron.tam@lge.com>

Commit:
rework HorizontalGridListImageItem to support right-aligned icon

This use case required that Overlay.Selection be applied to the Item
rather than Image so the icon could be position anywhere over the item.

Issue: ENYO-1641
Enyo-DCO-1.1-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

aarontam committed:
ENYO-1641: Small tweak for effective image size.

Enyo-DCO-1.1-Signed-off-by: Aaron Tam <aaron.tam@lge.com>
